### ISSUE TITLE
Circuit "Configuration" Data

### ISSUE THREAD

**Issue by dan-fritchman** [BWRC-AMS-ML-Discovery/AutoCkt#4]
There's one project-level YAML-format configuration file, which looks like so:

```yaml
database_dir: "Autockt/checkpoint/db/two_stage_opamp_ngspice"
dsn_netlist: "eval_engines/ngspice/ngspice_inputs/netlist/two_stage_opamp.cir"
num_process: 1

params:
  mp1:  !!python/tuple [1, 100, 1]
  mn1:  !!python/tuple [1, 100, 1]
  mp3:  !!python/tuple [1, 100, 1]
  mn3:  !!python/tuple [1, 100, 1]
  mn4:  !!python/tuple [1, 100, 1]
  mn5:  !!python/tuple [1, 100, 1]
  cc:   !!python/tuple [!!float 0.1e-12, !!float 10.0e-12, !!float 0.1e-12]

#Every specification output will be normlaized in the order of (gain, ibias, pm, ugbw)
normalize: !!python/tuple [350, 0.001, 60, 950000.0]

#during training, at reset a specification will be chosen at random between [min,max,#]
target_specs:
  gain_min: !!python/tuple [200,400]
  ugbw_min: !!python/tuple [!!float 1.0e6, !!float 2.5e7]
  phm_min: !!python/tuple [60,60.0000001]
  ibias_max: !!python/tuple [0.0001, 0.01]
```

So:

* That has 17 non-empty, non-comment lines.
* Of them, 13 require the !!python modifier.

In other words: that ain't YAML, it's just Python code with worse syntax.  
Just make it code.

We should add: the Python language has some helpful features which were not available when (some of) AutoCkt was written. The dataclasses added in Python 3.7 are particularly helpful for something like this. (They'd probably be an equal number of lines-of-code, and probably fewer characters on those lines.)

**Comment by dan-fritchman** [comment]
Assignee: @GuoruiCui

**Comment by dan-fritchman** [comment]
(Conversation started on Slack) - 

Are the three parameters in each of those (very silly) `!!python/tuple`s (min, max, resolution)?
I.e. the first six only take on integer values, and cc only takes on multiples of 1e-12?

Then… what are the values of the `normalize` and `target_specs` tuples? Don’t really follow those.

Assuming we are trying to specify things like `(min, max, step)`, just (a) make this code, and (b) make something like 

```python
from pydantic.dataclasses import dataclass 

Number = Union[int, float]

@dataclass 
class ParamSpec:
  min: Number
  max: Number 
  step: Number 

@dataclass 
class ParamSpecs: 
  mp1: ParamSpec
  # ...
  cc: ParamSpec

@dataclass 
class Normalize: 
  # ... whatever those mean 

@dataclass 
class Range:
  min: Number 
  max: Number 

@dataclass 
class TargetSpecs:
  ugbw: Range
  # ...
```

**Comment by kingh0730** [comment]
- `target_specs` specify the min and max of a target spec (which the RL agent tries to change the params to achieve) during one episode.
- `normalize` is just a specific spec. It's used as ML magic (no particular scientific significance) to calculate observations and rewards in RL.

@wayne-wang-1119

**Comment by kingh0730** [comment]
Can write some Constraints and Objectives (in the optimization sense) dataclasses, or maybe even use existing optimization libraries?

**Comment by dan-fritchman** [comment]
OK that makes more sense.  
So maybe we want something more like: 

```python
from pydantic.dataclasses import dataclass 

Number = Union[int, float]

@dataclass 
class ParamSpec:
  name: str # NEW! Parameter name
  min: Number
  max: Number 
  step: Number 

@dataclass 
class Params: 
  # either like List[Param] or Dict[str, Param] or similar

@dataclass 
class Spec: # <- THIS CHANGED
  name: str # Spec name
  min: Number 
  max: Number 
  # Notably, add the normalization constant as a feature of a `Spec`
  normalize: Number

@dataclass 
class Specs:
  # either like List[Spec] or Dict[str, Spec] or similar
```

Now the other part we talked about yesterday is specifying a cost/ reward/ "figure of merit" function. 
Recommend we: 

* Have a `CircuitOptimization` thing, which, maybe automatically from the types above, has: 
  * The input parameters that go into circuits-land (e.g. `AutoCktInput`) 
  * The output stuff coming back from circuits-land (e.g. `AutoCktOutput`)

Something like: 

```python
@dataclass 
class CircuitOptimization:
  # The stuff above, originally specified in YAML
  params: Params:
  specs: Specs

  # The "to and from circuit-stuff" types *could* be derived from `params` and `specs`:
  input_type: Type # Input to circuit-stuff
  output_type: Type # Outputs from circuit-stuff 
  # Initially - let's just write self-consistent ones, 
  # i.e. where the fields of `input_type` equal the `name`s of `params`, 
  # and the fields of `output_type` equal the `name`s of `specs`.

  # Reward function from circuit-outputs to float 
  reward: Callable[["Self.OutputType"], float] 
```

So `AutoCkt` would look like: 

```python
auto_ckt_opt = CircuitOptimization{
  input_type=AutoCktInput,
  output_type=AutoCktOutput,
  reward=that_reward_function_we_reviewed_yesterday
)
```

Maybe we'd also define some defaults in which the circuit-land largely computes `reward`, and the client-side `reward` function is just the identity function. E.g.: 

```python
def identity(r: float) -> float: 
  """ The identity function, serving as the default reward, if the circuit-server computes "goodness" """ 
 return r

@dataclass 
class CircuitOptimization:
  input_type: Type # Stays the same

  # These add defaults: 
  output_type: Type = float
  reward: Callable[["Self.OutputType"], float] = identity
```

**Comment by wayne-wang-1119** [comment]
partially done with the 1st part, now extremely generalizable dataclass on branch feature/dataclass, moving on to tweak reward

**Comment by kingh0730** [comment]
The way I understand it is this:
- Example client/server does ExampleOpAmp simulation
	- AutoCktInput -> AutoCktOutput (7 numbers -> 4 numbers)
- Ckt (current package name: `eval_engines`, might need to be renamed)
	- Things that your average ML researcher does not understand
	- exports:
		- ExampleOpAmpParams
		- ExampleOpAmpSpecs
		- ExampleOpAmpAction
		- ExampleOpAmpObservation
		- ExampleOpAmpReward
- Auto (current package name: `autockt`)
	- Things that your average circuit researcher does not understand
	- An RL lib that does what keerthana does, but without knowing what the circuit actually is
	- exports:
		- Trainer
		- Validator
		- Some `derive_XXX` functions, so that:
			- We can derive Action from Params if Action is not defined.
			- We can derive Observation from Params and Specs if Observation is not defined.
			- We can derive Reward from Specs (and Action?) if Reward is not defined.
- Shared
	- Auto and Ckt do not directly depend on each other, but they depend on Shared.
	- exports:
		- Params (a general type)
		- Specs (a general type)
		- Action (a general type)
		- Observation (a general type)
		- Reward (a general type)

**Comment by kingh0730** [comment]
Actually, Action and Observation should be ActionSpace and ObservationSpace.

**Comment by dan-fritchman** [comment]
I don't see how you could derive the reward function. 

And I would tweak the read from "does an ML researcher need to know what this is" to "which environment would need to run it". And the idea of having an ML-client-side reward function is that *it can be changed without needing cooperation from the circuit server*. And often will! There are many such related goals to be pursued. And if there *is* a specific, server-side reward function, that's fine too! The server can just return its single-number result, and the client can pass it unmodified through its own reward-function.

**Comment by kingh0730** [comment]
What I had in mind was somewhat different from what you meant on Tuesday. In my mind, Auto and Ckt are *both* client libraries.  And the actual server simulation lives in ExampleServer (which can be renamed) at the moment. So, Ckt uses ExampleClient to do OpAmp simulation, and Ckt creates additional RL-related things, like Reward, bounds on the Specs, etc.

The reason that Auto is separated from Ckt is that Auto handles training-related things, such as:
- The control flow in training
- How many CPU cores, and GPUs does my computer have
- Batch size, learning rate, etc.
- Where to store model folder

Supposed we want to train an agent that understands another type of circuit, then we can write:
- AnotherOpAmpParams
- AnotherOpAmpSpecs
- AnotherOpAmpAction
- AnotherOpAmpObservation
- AnotherOpAmpReward
And use Auto as is to train it.

**Comment by dan-fritchman** [comment]
Seems there is a competing version of this issue here: 
https://github.com/BWRC-AMS-ML-Discovery/BwrcAmsMlDiscovery/issues/24

This version has more commentary. We can move it to the other repo, but please close the other and keep this one going. 

This here: https://github.com/BWRC-AMS-ML-Discovery/BwrcAmsMlDiscovery/blob/main/AutoCkt/Ckt/eval_engines/__init__.py

Let's make it something like: 

```python
from wherever import (
    CircuitOptimization,
    Params,
    ParamSpec,
    Specs,
    Spec,
    AutoCktInput,
    AutoCktOutput,
)


def the_old_reward(spec: "???", goal_spec: "???") -> float:
    """
    Reward: doesn't penalize for overshooting spec, is negative
    FIXME: figure out where to get the `lookup` and `specs_id` data from.
    """

    rel_specs = self.lookup(spec, goal_spec)
    pos_val = []
    reward = 0.0
    for i, rel_spec in enumerate(rel_specs):
        if self.specs_id[i] == "ibias_max":
            rel_spec = rel_spec * -1.0  # /10.0
        if rel_spec < 0:
            reward += rel_spec
            pos_val.append(0)
        else:
            pos_val.append(1)

    return reward if reward < -0.02 else 10


def the_new_reward(arg: AutoCktOutput) -> float:
    # FIXME: basically do the logic in `the_old_reward`
    # but take that `AutoCktOutput` as input instead of `spec` and `goal_spec`
    ...


CircuitOptimization(
    params=Params(
        [
            ParamSpec("mp1", 1, 100, 1),
            ParamSpec("mp3", 1, 100, 1),
            ParamSpec("mn1", 1, 100, 1),
            ParamSpec("mn3", 1, 100, 1),
            ParamSpec("mn4", 1, 100, 1),
            ParamSpec("mn5", 1, 100, 1),
            ParamSpec("cc", 0.1e-12, 10.0e-12, 0.1e-12),
        ]
    ),
    specs=Specs(
        [
            Spec("gain", 200, 400, 350),
            Spec(
                "ibias", 1.0e6, 2.5e7, 1.0e7
            ),  # Note there are a few discrepancies on `ibias` vs `ibias_max`
            Spec("phm", 60, 60.0000001, 60),
            Spec("ugbw", 0.0001, 0.01, 950000.0),
        ]
    ),
    input_type=AutoCktInput,
    output_type=AutoCktOutput,
    reward=the_new_reward,
)
```

**Comment by kingh0730** [comment]
Check this out:
https://github.com/BWRC-AMS-ML-Discovery/BwrcAmsMlDiscovery/blob/32c4d2a564bd5204efcecde829a00f8ccb12045f/AutoCkt/scripts/train_general.py

**Comment by kingh0730** [comment]
Closing this, because we reached a consensus on this issue. The implementation of details are in:
https://github.com/BWRC-AMS-ML-Discovery/BwrcAmsMlDiscovery/issues/25
https://github.com/BWRC-AMS-ML-Discovery/BwrcAmsMlDiscovery/issues/26
https://github.com/BWRC-AMS-ML-Discovery/BwrcAmsMlDiscovery/issues/27
https://github.com/BWRC-AMS-ML-Discovery/BwrcAmsMlDiscovery/issues/29
and beyond
